Mathlib/Topology/Algebra/RestrictedProduct.lean

```diff
@@ -0,0 +1,769 @@
+
+Similarly, if `S` is a principal filter, the restricted product `Πʳ i, [R i, A i]_[𝓟 s]`
+is the set of all `x : Π i, R i` such that `∀ j ∈ S, x j ∈ A j`. -/
+def RestrictedProduct (𝓕 : Filter ι) : Type _ := {x : Π i, R i // ∀ᶠ i in 𝓕, x i ∈ A i}
+
+open Batteries.ExtendedBinder
+
+/-- `Πʳ i, [R i, A i]_[𝓕]` is `RestrictedProduct R A 𝓕`. -/
+scoped[RestrictedProduct]
+notation3 "Πʳ "(...)", ""["r:(scoped R => R)", "a:(scoped A => A)"]_[" f "]" =>
+  RestrictedProduct r a f
+
+/-- `Πʳ i, [R i, A i]` is `RestrictedProduct R A cofinite`. -/
+scoped[RestrictedProduct]
+notation3"Πʳ "(...)", ""["r:(scoped R => R)", "a:(scoped A => A)"]" =>
+  RestrictedProduct r a cofinite
+
+namespace RestrictedProduct
+
+open scoped RestrictedProduct
+
+variable {𝓕 𝓖 : Filter ι} {S T : Set ι}
+
+instance : DFunLike (Πʳ i, [R i, A i]_[𝓕]) ι R where
+  coe x i := x.1 i
+  coe_injective' _ _ := Subtype.ext
+
+lemma range_coe :
+    range ((↑) : Πʳ i, [R i, A i]_[𝓕] → Π i, R i) = {x | ∀ᶠ i in 𝓕, x i ∈ A i} :=
+  Subtype.range_val_subtype
+
+lemma range_coe_principal :
+    range ((↑) : Πʳ i, [R i, A i]_[𝓟 S] → Π i, R i) = S.pi A :=
+  range_coe R A
+
+variable (𝓕) in
+/-- The *structure map* of the restricted product is the obvious inclusion from `Π i, A i`
+into `Πʳ i, [R i, A i]_[𝓕]`. -/
+def structureMap (x : Π i, A i) : Πʳ i, [R i, A i]_[𝓕] :=
+  ⟨fun i ↦ x i, .of_forall fun i ↦ (x i).2⟩
+
+/-- If `𝓕 ≤ 𝓖`, the restricted product `Πʳ i, [R i, A i]_[𝓖]` is naturally included in
+`Πʳ i, [R i, A i]_[𝓕]`. This is the corresponding map. -/
+def inclusion (h : 𝓕 ≤ 𝓖) (x : Πʳ i, [R i, A i]_[𝓖]) :
+    Πʳ i, [R i, A i]_[𝓕] :=
+  ⟨x, x.2.filter_mono h⟩
+
+variable (𝓕) in
+lemma inclusion_eq_id : inclusion R A (le_refl 𝓕) = id := rfl
+
+lemma exists_inclusion_eq_of_eventually (h : 𝓕 ≤ 𝓖) {x : Πʳ i, [R i, A i]_[𝓕]}
+    (hx𝓖 : ∀ᶠ i in 𝓖, x i ∈ A i) :
+    ∃ x' : Πʳ i, [R i, A i]_[𝓖], inclusion R A h x' = x :=
+  ⟨⟨x.1, hx𝓖⟩, rfl⟩
+
+lemma exists_structureMap_eq_of_forall {x : Πʳ i, [R i, A i]_[𝓕]}
+    (hx : ∀ i, x.1 i ∈ A i) :
+    ∃ x' : Π i, A i, structureMap R A 𝓕 x' = x :=
+  ⟨fun i ↦ ⟨x i, hx i⟩, rfl⟩
+
+lemma range_inclusion (h : 𝓕 ≤ 𝓖) :
+    Set.range (inclusion R A h) = {x | ∀ᶠ i in 𝓖, x i ∈ A i} :=
+  subset_antisymm (range_subset_iff.mpr fun x ↦ x.2)
+    (fun _ hx ↦ mem_range.mpr <| exists_inclusion_eq_of_eventually R A h hx)
+
+lemma range_structureMap :
+    Set.range (structureMap R A 𝓕) = {f | ∀ i, f.1 i ∈ A i} :=
+  subset_antisymm (range_subset_iff.mpr fun x i ↦ (x i).2)
+    (fun _ hx ↦ mem_range.mpr <| exists_structureMap_eq_of_forall R A hx)
+
+section Algebra
+/-!
+## Algebraic instances on restricted products
+
+In this section, we endow the restricted product with its algebraic instances.
+To avoid any unnecessary coercions, we use subobject classes for the subset `A i` of each `R i`.
+-/
+
+variable {S S' : ι → Type*} -- subobject types
+variable [Π i, SetLike (S i) (R i)] [Π i, SetLike (S' i) (R' i)]
+variable {A : Π i, S i} {A' : Π i, S' i}
```

Comment:
You made A have type `(i : ι) → Set (R i)` globally in this file on line 110 so now we have got shadowed `A`s in our local context (and same for A'). This is a neat trick though.

Reply:
I'll change name to avoid shadowing.